src/envoy/http/authn/authenticator_base.cc

```diff
@@ -71,17 +72,59 @@ void AuthenticatorBase::validateX509(
     done_callback(&payload, false);
   }
 
-  // TODO (lei-tang): Adding other attributes (i.e ip) to payload if desire.
+  // TODO (lei-tang): Adding other attributes (i.e ip) to payload if needed.
   done_callback(&payload, true);
 }
 
+// Build JWT attributes from JWT payload
+// TODO (lei-tang): Build more attributes/claims to payload if needed.
+void buildJwtAttributes(JwtPayload* payload) {
+  ::google::protobuf::Map< ::std::string, ::std::string>& claims =
+      *payload->mutable_claims();
+  // Build user
+  if (claims.count("iss") > 0 && claims.count("sub") > 0) {
+    payload->set_user(claims["iss"] + "/" + claims["sub"]);
+  }
+  // Build audience
+  // The aud field is taken from jwt-auth produced claim as is.
+  if (claims.count("aud") > 0) {
+    payload->add_audiences(claims["aud"]);
+  }
+  // Build authorized presenter (azp)
+  if (claims.count("azp") > 0) {
+    payload->set_presenter(claims["azp"]);
+  }
+}
+
 void AuthenticatorBase::validateJwt(
     const iaapi::Jwt&,
-    const AuthenticatorBase::MethodDoneCallback& done_callback) const {
+    const AuthenticatorBase::MethodDoneCallback& done_callback) {
   Payload payload;
-  // TODO (diemtvu/lei-tang): construct jwt_authenticator and call Verify;
-  // pass done_callback so that it would be trigger by jwt_authenticator.onDone.
-  done_callback(&payload, false);
+  Envoy::Http::HeaderMap& header = *filter_context()->headers();
+  ENVOY_LOG(debug, "{} the number of headers is {}", __func__, header.size());
+
+  bool ret = AuthnUtils::GetJWTPayloadFromHeaders(
+      header, LowerCaseString(kJwtHeaderKey),
+      payload.mutable_jwt()->mutable_claims());
+  if (!ret) {
+    ENVOY_LOG(debug,
+              "AuthenticatorBase: {} GetJWTPayloadFromHeaders() returns false.",
+              __func__);
+    done_callback(nullptr, false);
+  } else {
+    if (payload.mutable_jwt()->mutable_claims()->empty()) {
+      ENVOY_LOG(error, "AuthenticatorBase: {} there is no JWT claims.",
+                __func__);
+      done_callback(nullptr, false);
+      return;
+    }
+    ENVOY_LOG(debug, "AuthenticatorBase: {}(): a valid JWT is found.",
+              __func__);
+    buildJwtAttributes(payload.mutable_jwt());
```

Comment:
Do you need to fail if required claims are not there?

Reply:
JWT verification is handled by jwt-auth filter, so the required claims have been verified by jwt-auth filter. If the verified JWT is non-empty, validateJwt() regards the JWT has been verified and outputs the verified claims from the jwt-auth filter.